plugins/inputs/tacacs/tacacs.go

```diff
@@ -0,0 +1,169 @@
+	Password        config.Secret   `toml:"password"`
+	Secret          config.Secret   `toml:"secret"`
+	RequestAddr     string          `toml:"request_ip"`
+	ResponseTimeout config.Duration `toml:"response_timeout"`
+	Log             telegraf.Logger `toml:"-"`
+	clients         []tacplus.Client
+	authStart       tacplus.AuthenStart
+}
+
+//go:embed sample.conf
+var sampleConfig string
+
+func (t *Tacacs) SampleConfig() string {
+	return sampleConfig
+}
+
+func (t *Tacacs) Init() error {
+	if len(t.Servers) == 0 {
+		t.Servers = []string{"127.0.0.1:49"}
+	}
+
+	if t.Username.Empty() || t.Password.Empty() || t.Secret.Empty() {
+		return errors.New("empty credentials were provided (username, password or secret)")
+	}
+
+	if t.RequestAddr == "" {
+		t.RequestAddr = "127.0.0.1"
+	}
+	if net.ParseIP(t.RequestAddr) == nil {
+		return fmt.Errorf("invalid ip address provided for request_ip: %s", t.RequestAddr)
+	}
+
+	t.clients = make([]tacplus.Client, 0, len(t.Servers))
+	for _, server := range t.Servers {
+		t.clients = append(t.clients, tacplus.Client{
+			Addr:       server,
+			ConnConfig: tacplus.ConnConfig{},
+		})
+	}
+
+	t.authStart = tacplus.AuthenStart{
+		Action:        tacplus.AuthenActionLogin,
+		AuthenType:    tacplus.AuthenTypeASCII,
+		AuthenService: tacplus.AuthenServiceLogin,
+		PrivLvl:       1,
+		Port:          "heartbeat",
+		RemAddr:       t.RequestAddr,
+	}
+
+	return nil
+}
+
+func (t *Tacacs) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+
+	for idx := range t.clients {
+		wg.Add(1)
+		go func(client *tacplus.Client) {
+			defer wg.Done()
+			acc.AddError(t.pollServer(acc, client))
+		}(&t.clients[idx])
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (t *Tacacs) pollServer(acc telegraf.Accumulator, client *tacplus.Client) error {
+	// Create the fields for this metric
+	tags := map[string]string{"source": client.Addr}
+	fields := make(map[string]interface{})
+
+	secret, err := t.Secret.Get()
+	if err != nil {
+		return fmt.Errorf("getting secret failed: %w", err)
+	}
+	defer config.ReleaseSecret(secret)
+
+	client.ConnConfig.Secret = secret
+
+	username, err := t.Username.Get()
+	if err != nil {
+		return fmt.Errorf("getting username failed: %w", err)
+	}
+	defer config.ReleaseSecret(username)
+
+	password, err := t.Password.Get()
+	if err != nil {
+		return fmt.Errorf("getting password failed: %w", err)
+	}
+	defer config.ReleaseSecret(password)
+
+	ctx := context.Background()
+	if t.ResponseTimeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(t.ResponseTimeout))
+		defer cancel()
+	}
+
+	startTime := time.Now()
+	reply, session, err := client.SendAuthenStart(ctx, &t.authStart)
+	if err != nil {
+		return fmt.Errorf("error on new tacacs authentication start request to %s : %w", client.Addr, err)
+	}
+	defer session.Close()
+	if reply.Status != tacplus.AuthenStatusGetUser {
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
```

Comment:
Honestly I have hard time deciding whats best. For our use case how we use it - we are only interested in successful auth response times, if its not successful, the response time is irrelevant to us and warrants an alert creation (whether by bad response code, or by longer than allowed response time). But agree perhaps some users might find it useful for different use cases. Now its possible to check response_code==1 for successful only, so we can live with anything here.

Wondering if context.DeadlineExceeded err can be hit with timeout set to 0, because then we would be filling the response time with 0

Reply:
Yeah, I do think its expected to have a real response time as it might indicate something why authentication cannot continue maybe?

Reply:
` I'm not sure if the status code should be a tag. Looks more like it should be a field to me. Or is it to be expected it to always be the same for a given configuration?`  

It can change if the remote server goes down, users password changes on remote server etc. If all stays the same on both client and server sides, should be always the same value.

Reply:
So it is a monitoring status and thus should be a field..

Reply:
Im adding it as a field then, also outputting real response time if there is ANY response. Added notes to README so its clear whats happening.

Im also a bit tempted to replace the hard errors thrown by the client after SendAuthenStart with a custom response_code of "error" similar to "timeout", simply because they can happen often due to similar reasons to timeout (issues with network, even wrong SECRET causes `return fmt.Errorf("error on new tacacs authentication start request to %s : %w", client.Addr, err)` while wrong password and wrong username cause regular response.
So instead of current:
| Field value | Meaning                   | From          | responsetime_ms
| ----------- | ------------------------- | ------------- | ---------------
| 1           | AuthenStatusPass          | tacacs server | real value
| 2           | AuthenStatusFail          | tacacs server | real value
| 3           | AuthenStatusGetData       | tacacs server | real value
| 4           | AuthenStatusGetUser       | tacacs server | real value
| 5           | AuthenStatusGetPass       | tacacs server | real value
| 6           | AuthenStatusRestart       | tacacs server | real value
| 7           | AuthenStatusError         | tacacs server | real value
| 33          | AuthenStatusFollow        | tacacs server | real value
| timeout     | timeout                   | telegraf      | eq. to response_timeout

we would also have
| Field value | Meaning                   | From          | responsetime_ms
| ----------- | ------------------------- | ------------- | ---------------
| error | error | telegraf      | eq. to response_timeout

Not sure where we would output the actual error details, whether into warn telegraf log, or into some additional field that is either ok or the error message or similar

Reply:
Ouch, that would mean the type suddenly will change from integer to string. A lot of outputs will not be able to handle that..

I would suggest to use `-1` instead.

Reply:
Added const reverse translation function (since underlying dependency does not offer it), and returning human readable code string representation now.